Ruby Capstone project.

shadabwahidullah opened the pull request:
# In this milestone we have:

- Added classes and logic based on the UML diagram
- Created a UI for our project
- Added the utility to save data as JSON files
- Added a schema.sql file for data

## Used Technologies:

- Ruby
- Linters

wikitelmex1 committed:
comit

item class still in process

ahmedadel56 committed:
Add can_be-archived , move_to_archive methods to items class

shadabwahidullah committed:
Create music_album & Implement methods for it

shadabwahidullah committed:
Create genere & Implement it's methods

shadabwahidullah committed:
Add id for genere

shadabwahidullah committed:
Correted genere typo in item and music album

shadabwahidullah committed:
Resolve some linter's errors

shadabwahidullah committed:
Restructure intialize method

shadabwahidullah committed:
Resolve bugs of genere

shadabwahidullah committed:
Resolve bugs of music album

ahmedadel56 committed:
Add book and label classes and their methods

ahmedadel56 committed:
Remove comment on private

shadabwahidullah committed:
Add encapsulation for gerere and music album

shadabwahidullah committed:
Resolve linters' errors

ahmedadel56 committed:
Solve conflict

shadabwahidullah committed:
Remove id from initialize

shadabwahidullah committed:
Correct the spelling for genre & Add id and archived to item initialize method

shadabwahidullah committed:
Merge pull request #33 from Wikitelmex/feature-tm-2

Feature tm 2

shadabwahidullah committed:
Merge branch 'develop' into team-member-one

shadabwahidullah committed:
Update books initialize method params

shadabwahidullah committed:
Resolve linters' errors

shadabwahidullah committed:
Merge pull request #32 from Wikitelmex/team-member-one

Add requird tasks for your role

wikitelmex1 committed:
Game class

- add game class file
- create game class attributes
- add constructor
- add override method can be archived

shadabwahidullah committed:
Correct a typo in genre class

shadabwahidullah committed:
Add Initialization test for genre class

shadabwahidullah committed:
Add intialization test for genre with id

shadabwahidullah committed:
Add initialization tests for music album

shadabwahidullah committed:
Add id and acrchived in music album's intialization method

shadabwahidullah committed:
Resolve linters' errors

shadabwahidullah committed:
Correct genre typo in add_item method & Add reader for genre items

shadabwahidullah committed:
Add test for adding an item to genre

shadabwahidullah committed:
Resolve linters' errors

wikitelmex1 committed:
Summary

## add class author
- fit attributes to schema
- add methods add item, to_json

## add class game
- fit attributes to schema
- add methods to_json

wikitelmex1 committed:
Update item.rb

wikitelmex1 committed:
Update item.rb

wikitelmex1 committed:
Update sanbox.rb

- fix linters

wikitelmex1 committed:
Summary

- implement MVC architectural pattern for authors
- add controller for author
- add icontroller interface
- add author view
- update UI to call author controller

wikitelmex1 committed:
summary

- some changes debugged

wikitelmex1 committed:
Update main.rb

- fix linters

wikitelmex1 committed:
Merge pull request #35 from Wikitelmex/feature/tm1-game

Feature/tm3 game

shadabwahidullah committed:
Edit tests run script

shadabwahidullah committed:
Merge pull request #34 from Wikitelmex/testing-tm2

Testing Team member 2

ahmedadel56 committed:
Add book and label classes then Add their methods

ahmedadel56 committed:
Fix linter errors and Fix time formate

ahmedadel56 committed:
Merge pull request #36 from Wikitelmex/team-member-one

Add requird tasks for your role

wikitelmex1 committed:
authors data preserve

- add entity.rb file to write and load data from JSON file
- add to_array method in author class to convert any JSON object into an array of objects of <T> Class

- add tests to author class

wikitelmex1 committed:
Update author_spec.rb

ahmedadel56 committed:
Add tests for book class

ahmedadel56 committed:
Add tests for label class

shadabwahidullah committed:
"Add getter for genre items"

shadabwahidullah committed:
"Add genre actions & Implement its methods"

shadabwahidullah committed:
"Add genreView and Display message methods"

shadabwahidullah committed:
Add genre to main & uiapp

wikitelmex1 committed:
summary

last updates

shadabwahidullah committed:
Add accessor for publish date in item

shadabwahidullah committed:
Add music album controller and Implement read and create methods

shadabwahidullah committed:
Add music album to show on UI & Implement it's methods

shadabwahidullah committed:
Add and initialize music album in UiApp

shadabwahidullah committed:
Resolve linters' errors

ahmedadel56 committed:
Add optional changes

wikitelmex1 committed:
Generic loader

## summary
add an generic loader to entity, when you specify certain class, and it will MAP from an JSON object to the specified class

- delete method self.to_array from author class
- add method self.constructor pattern to author class
- modify load method on entity class to load generically

ahmedadel56 committed:
Add schema file for gener,author,label,item and book classes

ahmedadel56 committed:
fix tests errors

wikitelmex1 committed:
Update author_spec.rb

- delete test to_array methot, is not loger needed

wikitelmex1 committed:
Delete sanbox.rb

wikitelmex1 committed:
Update .gitignore

wikitelmex1 committed:
Merge branch 'feature/tm3-preserve-data' of https://github.com/Wikitelmex/Ruby-capstone into feature/tm3-preserve-data

ahmedadel56 committed:
Fix required changes

wikitelmex1 committed:
Merge pull request #39 from Wikitelmex/feature/tm3-preserve-data

Feature/tm3 preserve data

ahmedadel56 committed:
Merge pull request #37 from Wikitelmex/team-member-one

 Add unit tests for all implemented methods

ahmedadel56 committed:
Save and preserve data for labels

shadabwahidullah committed:
Merge branch 'develop' into ui-options-tm2

shadabwahidullah committed:
Merge pull request #38 from Wikitelmex/ui-options-tm2

UI options tm2

shadabwahidullah committed:
Add preseve functionality to genre

shadabwahidullah committed:
Add preserve data functionality to music_album

shadabwahidullah committed:
Resolve linters' errors

shadabwahidullah committed:
Merge pull request #41 from Wikitelmex/preserve-tm2

Preserve tm2

wikitelmex1 committed:
Merge branch 'develop' into tm1-preserve-data

wikitelmex1 committed:
Merge pull request #40 from Wikitelmex/tm1-preserve-data

Save and preserve data for labels

ahmedadel56 committed:
Add create book and list books functionalities

ahmedadel56 committed:
Fix display all of books white spaces

shadabwahidullah committed:
Merge branch 'develop' into tm1-preserve-data

shadabwahidullah committed:
Resolve some linters' errors

shadabwahidullah committed:
Resolve linters' errors

shadabwahidullah committed:
Add more tests for genre

shadabwahidullah committed:
Add some more tests for genre and music album

shadabwahidullah committed:
Merge pull request #45 from Wikitelmex/tm2-more-tests

Tm2 more tests

ahmedadel56 committed:
Finish preserve data task

ahmedadel56 committed:
Merge branch 'develop' into tm1-preserve-data

ahmedadel56 committed:
Add required changes

ahmedadel56 committed:
Merge pull request #44 from Wikitelmex/tm1-preserve-data

Ahmed Adel's tasks

wikitelmex1 committed:
Games

- add MVC for games
- add games to UI
- preserve data for games

wikitelmex1 committed:
Create authors.json

shadabwahidullah committed:
Merge pull request #46 from Wikitelmex/feature/games-branch-tm3

Feature/games branch tm3

mikethreels commented on model/game.rb:
```diff
@@ -0,0 +1,34 @@
+    @last_played_at = last_played_at
+  end
+
+  def can_be_archived?
+    today_time = Time.now
+    today_date = Date.new(today_time.year, today_time.month, today_time.day)
+    (today_date - Date.parse(@publish_date)) > 36_525 && (today_date - Date.parse(@last_played_at)) > 730
+  end
```
- [ ] for this method you should use the parent method instead of checking it again manually this would make it look something like:
```
super && (today_date - Date.parse(@last_played_at)) > 730
```

Wikitelmex replied:
add method to calcultate the years passed since current time:

´´´´
  def years_since_today(date1)
    today_time = Time.now
    today_date = Date.new(today_time.year, today_time.month, today_time.day)
    mydate1 = Date.parse(date1)
    ((today_date - mydate1) / 365.25).to_i
  end
´´´´

mikethreels commented on specs/author_spec.rb:
```diff
@@ -0,0 +1,40 @@
+require_relative '../model/author'
+require 'json'
+
+describe Author do
```
- [ ] there are no unit tests for the `game` and `item` classes, please create unit tests for this as well.

DeliceLydia replied:
Well addressed this

mikethreels commented on uiapp.rb:
```diff
@@ -1,26 +1,101 @@
     puts '1 - List all books'
-    puts '2 - List all labels (e.g. Gift, New)'
-    puts '3 - Add a book'
+    puts '2 - Create a book'
+    puts '3 - List all labels (e.g. Gift, New)'
+    puts '4 - Create Label'
+    puts '5 - List all genres'
+    puts '6 - Create genre'
+    puts '7 - List all music albums'
+    puts '8 - Create Music album'
+    puts '9 - List all Authors'
+    puts '10 - Create Author'
+    puts '11 - List all Games'
+    puts '12 - Create Game'
```
- [ ] according to the requirements the application should be able to:
- List all books
- List all music albums
- List of games
- List all genres (e.g 'Comedy', 'Thriller')
- List all labels (e.g. 'Gift', 'New')
- List all authors (e.g. 'Stephen King')
- Add a book
- Add a music album
- Add a game
- Exit
the idea behind this is that when adding a book you should put in an author for example and that's how the authors get created, this should be connected.

mikethreels reviewed (changes requested):
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._
- [x] Please create a video presentation of your project with the entire team.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

wikitelmex1 committed:
Create game_spec.rb

- testing Game object creational variants
- testing game to_json  method

Wikitelmex commented on model/author.rb:
```diff
@@ -0,0 +1,27 @@
+    %w[first_name last_name id]
+  end
+
+  def add_item(item)
+    @items << item unless @items.include?(item)
+  end
```
added author association in both ways
´item.author = self´

DeliceLydia replied:
solved this

ahmedadel56 committed:
Add required changes

Wikitelmex commented on model/item.rb:
```diff
@@ -0,0 +1,40 @@
+    genre.items.push(self) unless genre.items.include?(self)
+  end
+
+  def author=(author)
+    @author = author
+    author.add_item(self)
+  end
+
+  def label=(label)
+    @label = label
+    label.add_item(self)
+  end
```
fixed issues in author
´
  def author=(author)
    @author = author
    author.items.push(self) unless author.items.include?(self)
  end
´

DeliceLydia replied:
Well resolved this!

wikitelmex1 committed:
Summary

# Summary
## Team member 3 fit requirements doing the next actions:

- fix connections between author and items
- clean require pry
- add method, to calculate years from a date since today
- add item class tests
- add game class tests

ahmedadel56 committed:
Merge pull request #48 from Wikitelmex/tm1-preserve-data

Add required changes

wikitelmex1 committed:
Merge pull request #49 from Wikitelmex/feature/game-requirements

Feature/game requirements

wikitelmex1 committed:
Update schema.sql

- add game table to fit the schema

ahmedadel56 committed:
Update schema.sql

Add music_albume table

wikitelmex1 committed:
Summary

- fit UI to the requirements
- improve list all views

wikitelmex1 committed:
UI

- little improvement to UI

wikitelmex1 committed:
Readme

- add video to Readme.MD

wikitelmex1 committed:
Merge pull request #51 from Wikitelmex/feature/ui-fit

Final touches to fit requirements

DeliceLydia commented on model/label.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def add_item(item)
+    @items << item unless @items.include?(item)
+  end
```
Addressed this!

DeliceLydia reviewed (approved):
# Congratulations🎉🎉

## Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



### Keep Rocking!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard).. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
